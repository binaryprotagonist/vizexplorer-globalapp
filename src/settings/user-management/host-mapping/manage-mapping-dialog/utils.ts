import { ApolloCache, FetchResult } from "@apollo/client";
import { sortArray } from "../../../../view/utils";
import { UserDisplay } from "../../../../view/user/utils";
import { MappingSelectNativeHostFragment } from "./__generated__/host-code-select";
import {
  HostMappingUpdateMutation,
  HostMappingUpdateMutationVariables,
  UserDocument,
  UserQuery
} from "./__generated__/manage-mapping-dialog";
import { HostSelectOption } from "./types";

export function mappingAsOption(
  mapping: MappingSelectNativeHostFragment
): HostSelectOption {
  return {
    id: mapping.nativeHostId,
    name: UserDisplay.fullNameV2(mapping)
  };
}

export function buildMappingOptions(
  unmapped: MappingSelectNativeHostFragment[],
  selected?: HostSelectOption[]
) {
  const unmappedOptions = unmapped.map(mappingAsOption);
  const combined = selected ? unmappedOptions.concat(selected) : unmappedOptions;

  return sortArray(combined, true, (option) => option.name);
}

export function mappingUpdateVariables(
  userId: string,
  siteId: string,
  mappings: HostSelectOption[]
): HostMappingUpdateMutationVariables {
  return {
    input: { userId, siteId, nativeHostIds: mappings.map((mapping) => mapping.id) }
  };
}

export function mappingCacheUpdate(
  userId: string,
  cache: ApolloCache<HostMappingUpdateMutation>,
  { data }: FetchResult<HostMappingUpdateMutation>
) {
  if (!data?.pdHostMappingUpdate) return;

  const userQuery: UserQuery | null = cache.readQuery({
    query: UserDocument,
    variables: { id: userId }
  });
  if (!userQuery?.user) return;

  const newUser: UserQuery["user"] = {
    ...userQuery.user,
    pdHostMappings: data.pdHostMappingUpdate
  };

  cache.writeQuery({
    query: UserDocument,
    variables: { id: userId },
    data: {
      user: newUser
    }
  });
}
