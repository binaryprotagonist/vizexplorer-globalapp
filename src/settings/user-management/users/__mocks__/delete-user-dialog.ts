import { UserDeleteDocument } from "../__generated__/delete-user-dialog";

export function mockUserDelete(userId: string) {
  return {
    request: {
      query: UserDeleteDocument,
      variables: { id: userId }
    },
    result: {
      data: { userDelete: true }
    }
  };
}
