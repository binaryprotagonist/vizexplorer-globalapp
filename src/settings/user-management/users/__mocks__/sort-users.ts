import { OrgAccessLevel } from "generated-graphql";
import { UserManagementUserFragment } from "../__generated__/users";

const sras = { id: "sras", name: "Slot Reports" };
const sre = { id: "sre", name: "Slot Recommendation Engine" };

const admin = { id: "admin", name: "App Admin" };
const viewer = { id: "viewer", name: "App Viewer" };

const site1 = { id: "1", name: "Zebra" };
const site2 = { id: "2", name: "Apple" };
const site3 = { id: "3", name: "Kevin" };
const site4 = { id: "4", name: "Cat" };

const orgAdmin1: UserManagementUserFragment = {
  id: "7856",
  email: "Xeno.Lore@test.com",
  firstName: "Xeno",
  lastName: "Lore",
  phone: "+1 123 321",
  accessLevel: OrgAccessLevel.OrgAdmin,
  accessList: [
    {
      app: sras,
      role: admin,
      site: site1
    },
    {
      app: sras,
      role: admin,
      site: site2
    },
    {
      app: sras,
      role: admin,
      site: site3
    },
    {
      app: sras,
      role: admin,
      site: site4
    }
  ]
};

const propertyAdmin1: UserManagementUserFragment = {
  id: "987453",
  email: "Pingu.Craze@test.com",
  firstName: "Pingu",
  lastName: "Craze",
  phone: "+1 123 321",
  accessLevel: OrgAccessLevel.AppSpecific,
  accessList: [
    {
      app: sras,
      role: admin,
      site: site1
    },
    {
      app: sras,
      role: admin,
      site: site2
    }
  ]
};

const propertyAdmin2: UserManagementUserFragment = {
  id: "7283",
  email: "Ribbon.End@test.com",
  firstName: "Ribbon",
  lastName: "End",
  phone: "+1 123 321",
  accessLevel: OrgAccessLevel.AppSpecific,
  accessList: [
    {
      app: sras,
      role: admin,
      site: site2
    }
  ]
};

const propertyAdmin3: UserManagementUserFragment = {
  id: "465610",
  email: "Unicorn.Bow@test.com",
  firstName: "Unicorn",
  lastName: "Bow",
  phone: "+1 123 321",
  accessLevel: OrgAccessLevel.AppSpecific,
  accessList: [
    {
      app: sras,
      role: admin,
      site: site4
    }
  ]
};

const propertyAdmin4: UserManagementUserFragment = {
  id: "16875",
  email: "Unicorn.Bow@test.com",
  firstName: "Unicorn",
  lastName: "Bow",
  phone: "+1 123 321",
  accessLevel: OrgAccessLevel.AppSpecific,
  accessList: [
    {
      app: sre,
      role: admin,
      site: site1
    },
    {
      app: sre,
      role: admin,
      site: site2
    },
    {
      app: sre,
      role: admin,
      site: site3
    },
    {
      app: sre,
      role: admin,
      site: site4
    }
  ]
};

const propertyViewer1: UserManagementUserFragment = {
  id: "18674",
  email: "Erin.Woke@test.com",
  firstName: "Erin",
  lastName: "Woke",
  phone: "+1 123 321",
  accessLevel: OrgAccessLevel.AppSpecific,
  accessList: [
    {
      app: sras,
      role: viewer,
      site: site1
    }
  ]
};

const propertyViewer2: UserManagementUserFragment = {
  id: "17658",
  email: "Gale.Bush@test.com",
  firstName: "Gale",
  lastName: "Bush",
  phone: "+1 123 321",
  accessLevel: OrgAccessLevel.AppSpecific,
  accessList: [
    {
      app: sras,
      role: viewer,
      site: site2
    },
    {
      app: sre,
      role: viewer,
      site: site4
    }
  ]
};

const propertyViewer3: UserManagementUserFragment = {
  id: "87654",
  email: "Hollywood.Key@test.com",
  firstName: "Hollywood",
  lastName: "Key",
  phone: "+1 123 321",
  accessLevel: OrgAccessLevel.AppSpecific,
  accessList: [
    {
      app: sras,
      role: viewer,
      site: site3
    },
    {
      app: sras,
      role: viewer,
      site: site4
    }
  ]
};

const emptyPropertyAccess1: UserManagementUserFragment = {
  id: "5738",
  email: "Zues.Pony@test.com",
  firstName: "Zues",
  lastName: "Pony",
  phone: "+1 123 321",
  accessLevel: OrgAccessLevel.AppSpecific,
  accessList: []
};

const emptyEmailUser: UserManagementUserFragment = {
  id: "9176",
  email: "",
  firstName: "James",
  lastName: "Smith",
  phone: "+1 423 902",
  accessLevel: OrgAccessLevel.AppSpecific,
  accessList: []
};

const noAccessUser: UserManagementUserFragment = {
  id: "8345",
  email: "Robert.Williams@test.com",
  firstName: "Robert",
  lastName: "Williams",
  phone: "+1 922 308",
  accessLevel: OrgAccessLevel.NoAccess,
  accessList: []
};

export const mockUnsortedUsers = [
  orgAdmin1,
  propertyAdmin1,
  propertyAdmin2,
  propertyAdmin3,
  propertyAdmin4,
  propertyViewer1,
  propertyViewer2,
  propertyViewer3,
  propertyAdmin4,
  emptyPropertyAccess1,
  emptyEmailUser,
  noAccessUser
];
