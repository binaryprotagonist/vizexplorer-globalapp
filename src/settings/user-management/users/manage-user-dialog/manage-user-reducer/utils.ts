import { ReducerUser } from "./types";

export function createReducerUser(): ReducerUser {
  return {
    firstName: "",
    lastName: "",
    phone: "",
    email: "",
    password: "",
    accessLevel: null,
    accessList: []
  };
}
