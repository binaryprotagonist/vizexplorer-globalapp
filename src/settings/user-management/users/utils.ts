import { GaAccessListFragment } from "generated-graphql";
import { isNoAccessUser, isOrgAdmin } from "../../../view/user/utils";
import { UserManagementUserFragment } from "./__generated__/users";

export function sortUsers(
  curUser: UserManagementUserFragment,
  users: UserManagementUserFragment[]
) {
  const usersCopy = [...users];
  const sortedUsers = usersCopy.sort((a, b) => compareUsersByRole(a, b, curUser));
  return sortedUsers.map((user) => {
    const accessCopy = [...user.accessList];
    const sortedAccessList = accessCopy.sort(compareAccessList);
    return { ...user, accessList: sortedAccessList };
  });
}

function compareAccessList(a: GaAccessListFragment, b: GaAccessListFragment) {
  const accessA = `${a.app.name}${a.role.name}${a.site.name}`.toLowerCase();
  const accessB = `${b.app.name}${b.role.name}${b.site.name}`.toLowerCase();

  if (accessA > accessB) return 1;
  return accessA < accessB ? -1 : 0;
}

/**
 * Sort users
 * Current user > Org Admins > Users related to Current User by App/Site > Remaining users
 */
function compareUsersByRole(
  a: UserManagementUserFragment,
  b: UserManagementUserFragment,
  currentUser: UserManagementUserFragment
) {
  // Handle null email addresses
  if (!a.email && !b.email) return 0;
  if (!a.email) return 1; // a has null email, place it after b
  if (!b.email) return -1; // b has null email, place it after a

  // current user
  if (a.id === currentUser.id) return -1;
  if (b.id === currentUser.id) return 1;

  // org admins
  if (isOrgAdmin(a.accessLevel)) return -1;
  if (isOrgAdmin(b.accessLevel)) return 1;

  // "No Access" users to be sorted to the bottom
  if (isNoAccessUser(a.accessLevel)) return 1;
  if (isNoAccessUser(b.accessLevel)) return -1;

  // other role ordering from the perspective of an Org Admin doesn't matter
  if (isOrgAdmin(currentUser.accessLevel)) return 0;

  // users that share some permissions with the current user
  if (accessIntersects(a.accessList, currentUser.accessList)) return -1;
  if (accessIntersects(b.accessList, currentUser.accessList)) return 1;

  // default case
  return 0;
}

// does accessA share similar access with accessB based on appId and siteId
function accessIntersects(
  accessA: GaAccessListFragment[],
  accessB: GaAccessListFragment[]
) {
  return accessA.some((a) =>
    accessB.some((b) => a.app.id === b.app.id && a.site.id === b.site.id)
  );
}
