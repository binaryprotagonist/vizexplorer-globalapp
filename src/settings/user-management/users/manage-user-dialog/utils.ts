import { GaUserFragment, OrgAccessLevel } from "generated-graphql";
import { ReducerAccess, ReducerUser } from "./manage-user-reducer/types";
import {
  ManagedSitesByApp,
  PasswordDisplay,
  UpdateOtherUserProps,
  UpdateOwnUserProps,
  UserManagement
} from "./types";
import {
  UserDisplay,
  canUser,
  isAppAdmin,
  isOrgAdmin
} from "../../../../view/user/utils";
import { UserActionType } from "../../../../view/user/types";
import {
  ManageUserAccessAppFragment,
  ManageUserAccessSiteFragment,
  UserCreateV2Mutation,
  UserCreateV2MutationVariables,
  UserUpdateMutation,
  UserUpdateMutationVariables,
  UsersQuery
} from "./__generated__/manage-user-dialog";
import {
  UserManagementUserFragment,
  UserManagementUserFragmentDoc
} from "../__generated__/users";
import { ApolloCache, FetchResult } from "@apollo/client";

const MIN_PASSWORD_LENGTH = 8;

export function accessLevelOptions(userAccessLevel: OrgAccessLevel) {
  const options = [OrgAccessLevel.AppSpecific, OrgAccessLevel.NoAccess];

  if (isOrgAdmin(userAccessLevel)) {
    options.unshift(OrgAccessLevel.OrgAdmin);
  }

  return options;
}

export function buildAppOptions(
  currentRow: ReducerAccess,
  allRows: ReducerAccess[],
  allApps: ManageUserAccessAppFragment[],
  managedSitesByApp: ManagedSitesByApp[]
): ManageUserAccessAppFragment[] {
  return allApps.filter((app) => {
    // is this app currently selected?
    if (app.id === currentRow.app?.id) return true;

    // does the user have permission to admin the app?
    const managedSitesForApp = managedSitesByApp.find((m) => m.app.id === app.id);
    if (!managedSitesForApp) return false;

    // have all properties for this app been exhausted for both existing and new access rows?
    const existingAccessForApp = allRows.filter((a) => a.app?.id === app.id);
    const usedSites = existingAccessForApp.map((access) => access.site?.id);
    if (managedSitesForApp.sites.every((site) => usedSites.includes(site.id))) {
      return false;
    }

    return true;
  });
}

export function buildSiteOptions(
  currentRow: ReducerAccess,
  allRows: ReducerAccess[],
  allSites: ManageUserAccessSiteFragment[],
  managedSitesByApp: ManagedSitesByApp[]
): ManageUserAccessSiteFragment[] {
  // don't return any options if the user hasn't selected an app first
  if (!currentRow.app?.id) return [];

  const rowAppId = currentRow.app.id;
  const managedSitesForApp = managedSitesByApp.find((m) => m.app.id === rowAppId);
  const managedSiteIds = managedSitesForApp?.sites.map((s) => s.id) ?? [];

  // only return any selected site if the user doesn't have permission to manage the app
  if (!managedSitesForApp) {
    if (currentRow.site?.id) {
      const selectedSite = allSites.find((s) => s.id === currentRow.site!.id);
      return selectedSite ? [selectedSite] : [];
    }

    return [];
  }

  const usedSiteIds = allRows
    .filter((row) => row.site?.id && row.app?.id === rowAppId)
    .map((access) => access.site!.id);

  // all sites exhausted
  if (usedSiteIds.length === allSites.length) {
    if (currentRow.site?.id) {
      const selectedSite = allSites.find((s) => s.id === currentRow.site!.id);
      return selectedSite ? [selectedSite] : [];
    }

    return [];
  }

  return allSites.filter((site) => {
    // if the site is currently selected, include it regardless of permissions
    if (site.id === currentRow.site?.id) return true;

    // does the user have admin access to the site and is it not already selected?
    return managedSiteIds.includes(site.id) && !usedSiteIds.includes(site.id);
  });
}

export function buildRoleOptions(
  currentRow: ReducerAccess,
  apps: ManageUserAccessAppFragment[]
): ManageUserAccessAppFragment["roles"] {
  // don't return any options if the user hasn't selected an app and site first
  if (!currentRow.app?.id || !currentRow.site?.id) return [];

  // return all roles for the selected app
  const selectedApp = apps.find((app) => app.id === currentRow.app!.id);
  return selectedApp?.roles ?? [];
}

export function buildManagedSitesByApp(
  currentUser: GaUserFragment,
  apps: ManageUserAccessAppFragment[],
  sites: ManageUserAccessSiteFragment[]
): ManagedSitesByApp[] {
  const managedApps = apps.filter((app) => {
    return app.status.isValid && isAppAdmin(currentUser, app.id);
  });

  return managedApps.map<ManagedSitesByApp>((app) => ({
    app,
    sites: sites.filter((site) =>
      canUser(currentUser, {
        type: UserActionType.MANAGE_APP_PROPERTY,
        appId: app.id,
        siteId: site.id
      })
    )
  }));
}

export function canEditAccessRow(
  currentUser: GaUserFragment,
  access: ReducerAccess
): boolean {
  if (isOrgAdmin(currentUser.accessLevel)) {
    return true;
  }

  // is app/site administrator?
  if (access.app?.id && access.site?.id) {
    return canUser(currentUser, {
      type: UserActionType.MANAGE_APP_PROPERTY,
      appId: access.app.id,
      siteId: access.site.id
    });
  }

  // is app administrator?
  if (access.app?.id) {
    return isAppAdmin(currentUser, access.app.id);
  }

  // no access defined, allow editing
  return true;
}

export function isFormComplete(
  formState: ReducerUser,
  // the access level of the user being edited before any changes to the form occurred, if applicable
  originalAccessLevel?: OrgAccessLevel
): boolean {
  const { firstName, lastName, phone, email, accessLevel, accessList } =
    trimUserWhitespace(formState);

  if (accessLevel === OrgAccessLevel.NoAccess) {
    return !!firstName && !!lastName;
  }

  const validPassword = isPasswordValid(formState, originalAccessLevel);
  const validAccess =
    accessLevel !== OrgAccessLevel.AppSpecific || isAccessListValid(accessList);

  return (
    validPassword &&
    validAccess &&
    !!accessLevel &&
    !!firstName &&
    !!lastName &&
    !!phone &&
    !!email
  );
}

function isAccessListValid(access: ReducerAccess[]): boolean {
  if (!access.length) return false;

  if (access.length === 1) {
    return isAccessComplete(access[0]);
  }

  // ignore empty access rows. only if some selection is made should it be validated
  return access.filter((row) => !isEmptyAccess(row)).every(isAccessComplete);
}

function isAccessComplete(access: ReducerAccess): boolean {
  return !!access.app?.id && !!access.site?.id && !!access.role?.id;
}

function isEmptyAccess(access: ReducerAccess): boolean {
  return !access.app?.id && !access.site?.id && !access.role?.id;
}

function isPasswordValid(
  formState: ReducerUser,
  originalAccessLevel?: OrgAccessLevel
): boolean {
  if (!formState.accessLevel) return false;

  const passwordDisplayType = passwordDisplay(formState.accessLevel, originalAccessLevel);
  switch (passwordDisplayType) {
    case "none":
      return formState.password === "";
    case "update":
      return !formState.password || formState.password.length >= MIN_PASSWORD_LENGTH;
    case "new":
      return formState.password.length >= MIN_PASSWORD_LENGTH;
  }
}

export function userToCreateUserV2Input(
  formState: ReducerUser
): UserCreateV2MutationVariables["input"] {
  const { firstName, lastName, phone, email, password, accessLevel, accessList } =
    trimUserWhitespace(formState);

  if (accessLevel === OrgAccessLevel.NoAccess) {
    return {
      firstName,
      lastName,
      phone,
      email: email ? email : null,
      accessLevel: accessLevel
    };
  }

  const accessForLevel = accessLevel === OrgAccessLevel.AppSpecific ? accessList : [];
  const completeAccess = accessForLevel.filter(isAccessComplete);

  return {
    firstName,
    lastName,
    phone,
    email,
    password,
    accessLevel: accessLevel!,
    accessList: completeAccess.map((access) => ({
      appId: access.app!.id!,
      siteId: access.site!.id!,
      roleId: access.role!.id!
    }))
  };
}

export function userToUserUpdateInput(
  currentUser: GaUserFragment,
  editUserId: string,
  formState: ReducerUser,
  userManagement: UpdateOtherUserProps | UpdateOwnUserProps
): UserUpdateMutationVariables["input"] {
  if (userManagement.type === "update-own-user") {
    return userToUserProfileUpdateInput(editUserId, formState);
  }

  if (isOrgAdmin(currentUser.accessLevel)) {
    return orgAdminUserToUserUpdateInput(editUserId, formState);
  } else {
    return appAdminUserToUserUpdateInput(editUserId, formState.accessList);
  }
}

function orgAdminUserToUserUpdateInput(
  userId: string,
  formState: ReducerUser
): UserUpdateMutationVariables["input"] {
  const { firstName, lastName, phone, email, password, accessLevel, accessList } =
    trimUserWhitespace(formState);

  if (accessLevel === OrgAccessLevel.NoAccess) {
    return {
      userId,
      firstName,
      lastName,
      phone,
      email,
      accessLevel: accessLevel
    };
  }

  const accessForLevel = accessLevel === OrgAccessLevel.AppSpecific ? accessList : [];
  const completeAccess = accessForLevel.filter(isAccessComplete);

  return {
    userId,
    firstName,
    lastName,
    phone,
    email,
    password: password !== "" ? password : null,
    accessLevel: accessLevel!,
    accessList: completeAccess.map((access) => ({
      appId: access.app!.id!,
      siteId: access.site!.id!,
      roleId: access.role!.id!
    }))
  };
}

function appAdminUserToUserUpdateInput(
  userId: string,
  updatedAccessList: ReducerUser["accessList"]
): UserUpdateMutationVariables["input"] {
  const completeAccess = updatedAccessList.filter(isAccessComplete);

  return {
    userId,
    accessList: completeAccess.map((access) => ({
      appId: access.app!.id,
      siteId: access.site!.id,
      roleId: access.role!.id
    }))
  };
}

function userToUserProfileUpdateInput(
  userId: string,
  formState: ReducerUser
): UserUpdateMutationVariables["input"] {
  const { firstName, lastName, phone, email, password } = trimUserWhitespace(formState);

  return {
    userId,
    firstName,
    lastName,
    phone,
    email,
    password: password !== "" ? password : null
  };
}

function trimUserWhitespace(user: ReducerUser): ReducerUser {
  return {
    firstName: user.firstName.trim(),
    lastName: user.lastName.trim(),
    phone: user.phone.trim(),
    email: user.email.trim(),
    password: user.password.trim(),
    accessLevel: user.accessLevel,
    accessList: user.accessList
  };
}

export function isNameTaken(user: ReducerUser, takenNames: string[]): boolean {
  if (user.accessLevel !== OrgAccessLevel.NoAccess) {
    return false;
  }

  const usersName = UserDisplay.fullName(user.firstName.trim(), user.lastName.trim());
  return takenNames.includes(usersName);
}

export function userManagementUserAsReducerUser(
  user: UserManagementUserFragment
): ReducerUser {
  return {
    firstName: user.firstName,
    lastName: user.lastName,
    accessLevel: user.accessLevel,
    email: user.email,
    phone: user.phone,
    accessList: user.accessList,
    password: ""
  };
}

export function noAccessUsersNames(
  users: NonNullable<UsersQuery["users"]>,
  excludeUserIds: string[]
): string[] {
  return users
    .filter((user) => {
      if (excludeUserIds.includes(user.id)) {
        return false;
      }

      return user.accessLevel === OrgAccessLevel.NoAccess;
    })
    .map(UserDisplay.fullNameV2);
}

// control whether we need to display `Change password` button, show `Password` field, or display no field
export function passwordDisplay(
  selectedAccessLevel: OrgAccessLevel,
  originalAccessLevel?: OrgAccessLevel
): PasswordDisplay {
  if (selectedAccessLevel === OrgAccessLevel.NoAccess) return "none";

  return !originalAccessLevel || originalAccessLevel === OrgAccessLevel.NoAccess
    ? "new"
    : "update";
}

export function requireEmailValidation(formState: ReducerUser, originalEmail?: string) {
  if (originalEmail && formState.email === originalEmail) {
    return false;
  }

  return formState.accessLevel !== OrgAccessLevel.NoAccess || !!formState.email;
}

export function hasFormChanged(
  formState: ReducerUser,
  originalUser: UserManagementUserFragment
): boolean {
  const trimmedFormState = trimUserWhitespace(formState);
  const trimmedOriginalState = trimUserWhitespace(
    userManagementUserAsReducerUser(originalUser)
  );

  if (trimmedFormState.firstName !== trimmedOriginalState.firstName) return true;
  if (trimmedFormState.lastName !== trimmedOriginalState.lastName) return true;
  if (trimmedFormState.phone !== trimmedOriginalState.phone) return true;
  if (trimmedFormState.email !== trimmedOriginalState.email) return true;
  if (trimmedFormState.accessLevel !== trimmedOriginalState.accessLevel) return true;
  if (trimmedFormState.password.length >= MIN_PASSWORD_LENGTH) return true;

  return hasAccessChanged(trimmedFormState.accessList, trimmedOriginalState.accessList);
}

function hasAccessChanged(
  newAccess: ReducerAccess[],
  originalAccess: ReducerAccess[]
): boolean {
  const completeAccess = newAccess.filter(isAccessComplete);
  if (completeAccess.length !== originalAccess.length) {
    return true;
  }

  return !completeAccess.every((newA) => {
    return originalAccess.some((originalA) => {
      return (
        newA.app?.id === originalA.app?.id &&
        newA.site?.id === originalA.site?.id &&
        newA.role?.id === originalA.role?.id
      );
    });
  });
}

export function showAccessLevel(userManagementType: UserManagement["type"]): boolean {
  return userManagementType !== "update-own-user";
}

export function showAccessList(
  userManagementType: UserManagement["type"],
  accessLevel: OrgAccessLevel | null
): boolean {
  return (
    userManagementType !== "update-own-user" && accessLevel === OrgAccessLevel.AppSpecific
  );
}

export function userCreateCacheUpdate(
  cache: ApolloCache<UserCreateV2Mutation>,
  { data }: FetchResult<UserCreateV2Mutation>
) {
  if (!data?.userCreateV2) return;

  cache.modify({
    fields: {
      users(existing = []) {
        const newUser = cache.writeFragment({
          data: data.userCreateV2,
          fragment: UserManagementUserFragmentDoc,
          fragmentName: "UserManagementUser"
        });
        return [...existing, newUser];
      }
    }
  });
}

export function userUpdateCacheUpdate(
  user: UserManagementUserFragment,
  cache: ApolloCache<UserUpdateMutation>,
  { data }: FetchResult<UserUpdateMutation>
) {
  // If the userId hasn't changed, cache will be automatically updated
  if (!data?.userUpdate || data.userUpdate.id === user.id) return;

  cache.evict({ id: cache.identify(user) });
  cache.gc();

  // UserID change occurs when changing a user between No Access and some Access. The user is recreated in BE, with a new ID
  // If the userId changed, remove the old user and insert a new user record
  cache.modify({
    fields: {
      users(existing = []) {
        const newUser = cache.writeFragment({
          data: data.userUpdate,
          fragment: UserManagementUserFragmentDoc,
          fragmentName: "UserManagementUser"
        });
        return [...existing, newUser];
      }
    }
  });
}
