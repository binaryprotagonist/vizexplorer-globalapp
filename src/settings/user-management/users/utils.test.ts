import { mockUnsortedUsers } from "./__mocks__/sort-users";
import { sortUsers } from "./utils";

describe("User Management Users Utils", () => {
  describe("sortUsers", () => {
    it("sorts users", () => {
      expect(sortUsers(mockUnsortedUsers[1], mockUnsortedUsers)).toMatchSnapshot();
    });
  });
});
