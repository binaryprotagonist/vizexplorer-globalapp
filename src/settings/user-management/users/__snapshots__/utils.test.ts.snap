// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Management Users Utils sortUsers sorts users 1`] = `
[
  {
    "accessLevel": "APP_SPECIFIC",
    "accessList": [
      {
        "app": {
          "id": "sras",
          "name": "Slot Reports",
        },
        "role": {
          "id": "admin",
          "name": "App Admin",
        },
        "site": {
          "id": "2",
          "name": "Apple",
        },
      },
      {
        "app": {
          "id": "sras",
          "name": "Slot Reports",
        },
        "role": {
          "id": "admin",
          "name": "App Admin",
        },
        "site": {
          "id": "1",
          "name": "Zebra",
        },
      },
    ],
    "email": "Pingu.Craze@test.com",
    "firstName": "Pingu",
    "id": "987453",
    "lastName": "Craze",
    "phone": "+1 123 321",
  },
  {
    "accessLevel": "ORG_ADMIN",
    "accessList": [
      {
        "app": {
          "id": "sras",
          "name": "Slot Reports",
        },
        "role": {
          "id": "admin",
          "name": "App Admin",
        },
        "site": {
          "id": "2",
          "name": "Apple",
        },
      },
      {
        "app": {
          "id": "sras",
          "name": "Slot Reports",
        },
        "role": {
          "id": "admin",
          "name": "App Admin",
        },
        "site": {
          "id": "4",
          "name": "Cat",
        },
      },
      {
        "app": {
          "id": "sras",
          "name": "Slot Reports",
        },
        "role": {
          "id": "admin",
          "name": "App Admin",
        },
        "site": {
          "id": "3",
          "name": "Kevin",
        },
      },
      {
        "app": {
          "id": "sras",
          "name": "Slot Reports",
        },
        "role": {
          "id": "admin",
          "name": "App Admin",
        },
        "site": {
          "id": "1",
          "name": "Zebra",
        },
      },
    ],
    "email": "Xeno.Lore@test.com",
    "firstName": "Xeno",
    "id": "7856",
    "lastName": "Lore",
    "phone": "+1 123 321",
  },
  {
    "accessLevel": "APP_SPECIFIC",
    "accessList": [
      {
        "app": {
          "id": "sre",
          "name": "Slot Recommendation Engine",
        },
        "role": {
          "id": "viewer",
          "name": "App Viewer",
        },
        "site": {
          "id": "4",
          "name": "Cat",
        },
      },
      {
        "app": {
          "id": "sras",
          "name": "Slot Reports",
        },
        "role": {
          "id": "viewer",
          "name": "App Viewer",
        },
        "site": {
          "id": "2",
          "name": "Apple",
        },
      },
    ],
    "email": "Gale.Bush@test.com",
    "firstName": "Gale",
    "id": "17658",
    "lastName": "Bush",
    "phone": "+1 123 321",
  },
  {
    "accessLevel": "APP_SPECIFIC",
    "accessList": [
      {
        "app": {
          "id": "sras",
          "name": "Slot Reports",
        },
        "role": {
          "id": "viewer",
          "name": "App Viewer",
        },
        "site": {
          "id": "1",
          "name": "Zebra",
        },
      },
    ],
    "email": "Erin.Woke@test.com",
    "firstName": "Erin",
    "id": "18674",
    "lastName": "Woke",
    "phone": "+1 123 321",
  },
  {
    "accessLevel": "APP_SPECIFIC",
    "accessList": [
      {
        "app": {
          "id": "sras",
          "name": "Slot Reports",
        },
        "role": {
          "id": "admin",
          "name": "App Admin",
        },
        "site": {
          "id": "2",
          "name": "Apple",
        },
      },
    ],
    "email": "Ribbon.End@test.com",
    "firstName": "Ribbon",
    "id": "7283",
    "lastName": "End",
    "phone": "+1 123 321",
  },
  {
    "accessLevel": "APP_SPECIFIC",
    "accessList": [
      {
        "app": {
          "id": "sras",
          "name": "Slot Reports",
        },
        "role": {
          "id": "admin",
          "name": "App Admin",
        },
        "site": {
          "id": "4",
          "name": "Cat",
        },
      },
    ],
    "email": "Unicorn.Bow@test.com",
    "firstName": "Unicorn",
    "id": "465610",
    "lastName": "Bow",
    "phone": "+1 123 321",
  },
  {
    "accessLevel": "APP_SPECIFIC",
    "accessList": [
      {
        "app": {
          "id": "sre",
          "name": "Slot Recommendation Engine",
        },
        "role": {
          "id": "admin",
          "name": "App Admin",
        },
        "site": {
          "id": "2",
          "name": "Apple",
        },
      },
      {
        "app": {
          "id": "sre",
          "name": "Slot Recommendation Engine",
        },
        "role": {
          "id": "admin",
          "name": "App Admin",
        },
        "site": {
          "id": "4",
          "name": "Cat",
        },
      },
      {
        "app": {
          "id": "sre",
          "name": "Slot Recommendation Engine",
        },
        "role": {
          "id": "admin",
          "name": "App Admin",
        },
        "site": {
          "id": "3",
          "name": "Kevin",
        },
      },
      {
        "app": {
          "id": "sre",
          "name": "Slot Recommendation Engine",
        },
        "role": {
          "id": "admin",
          "name": "App Admin",
        },
        "site": {
          "id": "1",
          "name": "Zebra",
        },
      },
    ],
    "email": "Unicorn.Bow@test.com",
    "firstName": "Unicorn",
    "id": "16875",
    "lastName": "Bow",
    "phone": "+1 123 321",
  },
  {
    "accessLevel": "APP_SPECIFIC",
    "accessList": [
      {
        "app": {
          "id": "sras",
          "name": "Slot Reports",
        },
        "role": {
          "id": "viewer",
          "name": "App Viewer",
        },
        "site": {
          "id": "4",
          "name": "Cat",
        },
      },
      {
        "app": {
          "id": "sras",
          "name": "Slot Reports",
        },
        "role": {
          "id": "viewer",
          "name": "App Viewer",
        },
        "site": {
          "id": "3",
          "name": "Kevin",
        },
      },
    ],
    "email": "Hollywood.Key@test.com",
    "firstName": "Hollywood",
    "id": "87654",
    "lastName": "Key",
    "phone": "+1 123 321",
  },
  {
    "accessLevel": "APP_SPECIFIC",
    "accessList": [
      {
        "app": {
          "id": "sre",
          "name": "Slot Recommendation Engine",
        },
        "role": {
          "id": "admin",
          "name": "App Admin",
        },
        "site": {
          "id": "2",
          "name": "Apple",
        },
      },
      {
        "app": {
          "id": "sre",
          "name": "Slot Recommendation Engine",
        },
        "role": {
          "id": "admin",
          "name": "App Admin",
        },
        "site": {
          "id": "4",
          "name": "Cat",
        },
      },
      {
        "app": {
          "id": "sre",
          "name": "Slot Recommendation Engine",
        },
        "role": {
          "id": "admin",
          "name": "App Admin",
        },
        "site": {
          "id": "3",
          "name": "Kevin",
        },
      },
      {
        "app": {
          "id": "sre",
          "name": "Slot Recommendation Engine",
        },
        "role": {
          "id": "admin",
          "name": "App Admin",
        },
        "site": {
          "id": "1",
          "name": "Zebra",
        },
      },
    ],
    "email": "Unicorn.Bow@test.com",
    "firstName": "Unicorn",
    "id": "16875",
    "lastName": "Bow",
    "phone": "+1 123 321",
  },
  {
    "accessLevel": "APP_SPECIFIC",
    "accessList": [],
    "email": "Zues.Pony@test.com",
    "firstName": "Zues",
    "id": "5738",
    "lastName": "Pony",
    "phone": "+1 123 321",
  },
  {
    "accessLevel": "NO_ACCESS",
    "accessList": [],
    "email": "Robert.Williams@test.com",
    "firstName": "Robert",
    "id": "8345",
    "lastName": "Williams",
    "phone": "+1 922 308",
  },
  {
    "accessLevel": "APP_SPECIFIC",
    "accessList": [],
    "email": "",
    "firstName": "James",
    "id": "9176",
    "lastName": "Smith",
    "phone": "+1 423 902",
  },
]
`;
